src/main/java/org/mskcc/cmo/metadb/service/impl/MessageHandlingServiceImpl.java

```diff
@@ -76,34 +79,56 @@ public void run() {
-                            continue;
-                        }
+                    String requestJson = cmoLabelGeneratorQueue.poll(100, TimeUnit.MILLISECONDS);
+                    if (requestJson != null && requestJson.length() > 10) {
                         // skip request if there are no samples to generate cmo labels for
-                        if (request.getSamples().isEmpty()) {
-                            requestStatusLogger.logRequestStatus(request.getRequestJson(),
+                        List<SampleMetadata> samples = getSamplesFromRequestJson(requestJson);
+                        if (samples.isEmpty()) {
+                            requestStatusLogger.logRequestStatus(requestJson,
                                     RequestStatusLogger.StatusType.REQUEST_WITH_MISSING_SAMPLES);
                             continue;
                         }
+                        Map<String, List<SampleMetadata>> patientSamplesMap = new HashMap<>();
+
                         // generate cmo labels for each sample metadata
                         List<SampleMetadata> updatedSamples = new ArrayList<>();
-                        for (SampleMetadata sample : request.getSamples()) {
-                            // generate cmo label
-                            String sampleCmoLabel = cmoLabelGeneratorService.generateCmoSampleLabel(sample);
+                        for (SampleMetadata sample : samples) {
+                            // get or request existing patient samples and update patient sample mapping
+                            if (!patientSamplesMap.containsKey(sample.getCmoPatientId())) {
+                                Message reply = messagingGateway.request(PATIENT_SAMPLES_REQUEST_TOPIC,
+                                        sample.getCmoPatientId());
+                                String replyData = mapper.readValue(
+                                        new String(reply.getData(), StandardCharsets.UTF_8),
+                                        String.class);
+                                patientSamplesMap.put(sample.getCmoPatientId(),
+                                        Arrays.asList(mapper.convertValue(replyData,
+                                                SampleMetadata[].class)));
+                            }
+                            List<SampleMetadata> existingSamples =
+                                    patientSamplesMap.get(sample.getCmoPatientId());
+
+                            // TODO resolve any issues that arise with errors in generating cmo label
+                            String sampleCmoLabel = cmoLabelGeneratorService.generateCmoSampleLabel(sample,
+                                    existingSamples);
                             sample.setCmoSampleName(sampleCmoLabel);
                             updatedSamples.add(sample);
+                            // update patient sample map also
+                            existingSamples.add(sample);
```

Comment:
Is it possible that the sample already existed in MetaDB and was returned by messagingGateway.request(PATIENT_SAMPLES_REQUEST_TOPIC,...) ?  I guess this would mean a sample existed in two different IGO requests.  In any event, colud this call be duplicating the sample in existingSamples?

Reply:
Should we check for whether a sample already exists by the generated sample label?

I didnt see code that was doing that specifically in the LimsRest repo but they might have also addressed issues like that manually, I'm not sure.